### PR TITLE
i.landsat.qa: qa bands from landsat collection 1and 2

### PR DESCRIPTION
**Pull request by ninsbl**
A first rough attempt at fixing #441 with a more general module.
Should be easy to add support for collection 2.
Needs still a bit more work esp. on documentation...

### CODE REVIEW

**Comment by ninsbl** [comment]
This should work and also support collection 2 QA bands. Once that is the default in i.landsat the answer for collection in this module can be adjusted.

Contributions to updating the manual are welcome...

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py [hunk 1]
```diff
@@ -0,0 +1,362 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+ MODULE:       i.landsat8.qa
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py:5]:
```suggestion
 MODULE:       i.landsat.qa
```

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+#% keyword: bitpattern
+#% keyword: mask
+#% keyword: landsat
+#%End
+
+#%option G_OPT_F_OUTPUT
+#% description: Output file with reclass rules
+#% required: no
+#%end
+
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py:34]: do we need 2 new lines here?

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+#% description: Landsat Collection (1 or 2)
+#% options: 1, 2
+#% answer: 1
+#% required : no
+#%end
+
+#%option
+#% key: sensor
+#% multiple: No
+#% description: Landsat Collection (1 or 2)
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py:48]:
```suggestion
#% description: Landsat sensor
```

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+#% options: Landsat 8 OLI, Landsat 8 OLI/TIRS, Landsat 1-5 MSS, Landsat 7 ETM+, Landsat 4-5 TM
+#% answer: 1
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.py:50]: is this ok? shall we answer with numbers?

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+<p>"Rigorous science applications seeking to optimize the value of pixels used 
+in a study will find QA bits useful as a first level indicator of certain 
+conditions. Otherwise, users are advised that this file contains information 
+that can be easily misinterpreted and it is not recommended for general use."</p>
+
+<h2>EXAMPLE</h2>
+
+<div class="code"><pre>
+#Create a cloud mask:
+i.landsat.qa --overwrite --verbose cloud="Maybe,Yes" \
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html:141]:
```suggestion
i.landsat.qa --overwrite --verbose cloud_confidence="Medium,High" \
```

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+<p>"Rigorous science applications seeking to optimize the value of pixels used 
+in a study will find QA bits useful as a first level indicator of certain 
+conditions. Otherwise, users are advised that this file contains information 
+that can be easily misinterpreted and it is not recommended for general use."</p>
+
+<h2>EXAMPLE</h2>
+
+<div class="code"><pre>
+#Create a cloud mask:
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html:140]:
```suggestion
# Create a cloud mask:
```

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+i.landsat.qa --overwrite --verbose cloud="Maybe,Yes" \
+    output=./Cloud_Mask_rules.txt
+r.reclass input=LC81980182015183LGN00_BQA \
+    output=LC81980182015183LGN00_Cloud_Mask rules=./Cloud_Mask_rules.txt
+
+#Create a water mask:
+i.landsat.qa --overwrite --verbose water="Maybe,Yes" \
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html:147]:
```suggestion
# Create a water mask (only for Collection 2):
i.landsat.qa --overwrite --verbose water="Yes" \
```

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+<a href="i.modis.qc.html">i.modis.qc</a>,
+<a href="r.bitpattern.html">r.bitpattern</a>,
+<a href="i.landsat.html">i.landsat</a>
+<a href="i.landsat8.swlst.html">i.landsat8.swlst</a>
+</em>
+
+<h2>REFERENCES</h2>
+
+<em><a href=http://landsat.usgs.gov/qualityband.php>http://landsat.usgs.gov/qualityband.php</a>
+<a href=https://landsat.usgs.gov/documents/LDCM_CVT_ADD.pdf>https://landsat.usgs.gov/documents/LDCM_CVT_ADD.pdf</a></em>
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html:168]: these 2 links yield 404

Maybe we can put links to Collection 1 and 2 product definitions:
- https://www.usgs.gov/media/files/landsat-collection-1-level-1-product-definition
- https://www.usgs.gov/core-science-systems/nli/landsat/landsat-collection-2-level-2-science-products

unfortunately they do not follow a common structure...

File: grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+in a study will find QA bits useful as a first level indicator of certain 
+conditions. Otherwise, users are advised that this file contains information 
+that can be easily misinterpreted and it is not recommended for general use."</p>
+
+<h2>EXAMPLE</h2>
+
+<div class="code"><pre>
+#Create a cloud mask:
+i.landsat.qa --overwrite --verbose cloud="Maybe,Yes" \
+    output=./Cloud_Mask_rules.txt
```
Comment (veroandreo) [inline grass7/imagery/i.landsat/i.landsat.qa/i.landsat.qa.html:142]: to make it clearer, we could add here `collection=1 sensor="Landsat 8 OLI/TIRS"` <-- is that right?

**Review by veroandreo — COMMENTED** [review]
This is great! I really like the idea of making it more general and also part of the toolset! Thanks for such a quick implementation! :D I'll think of a piece of text to say something about collections and levels in the description section of the manual. 

Also, I just checked the files changed now. I need to download the patch and apply it to my local copy of the toolset to really test it. I didn't because I was unsure about the `sensor` option.

**Comment by veroandreo** [comment]
Just tested locally with:
```
python i.landsat.qa.py cloud_confidence="Medium,High" output=/home/veroandreo/Cloud_Mask_rules.txt sensor="Landsat 8 OLI/TIRS"
r.reclass input=LC08_L1TP_229082_20200114_20200127_01_T1_BQA output=LC08_229082_20200114_Cloud_Mask rules=/home/veroandreo/Cloud_Mask_rules.txt --o
```
and it works! :) 

Indeed the option `sensor` left to default gives error, so now I confirm to myself that 1 as default answer there is a leftover of copy-pasting from option collection.

**Comment by veroandreo** [comment]
[landsatxplore](https://github.com/yannforget/landsatxplore) just got updated to support Landsat collection 2 data  (https://github.com/yannforget/landsatxplore/issues/29#issuecomment-785238702) and I'll update i.landsat.import in the next couple of days. This is a great chance then to complete i.landsat toolset :nerd_face:

**Comment by ninsbl** [comment]
Thanks for your feedback. I applied your suggestions with some additions (and todos). If everything works, this could be merged as far as I am concerned....

**Review by veroandreo — APPROVED** [review]
merge :)

**Pull request merged**